### PR TITLE
Default behaviour for workload bursting

### PR DESCRIPTION
**Pull request by meln1k**
Summary: when there are no region/zone related constraints, we should accept offers only from the local region

JIRA issues: MARATHON_EE-1695

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala [hunk 1]
```diff
@@ -67,16 +72,16 @@ class ActorOfferMatcherTest extends AkkaUnitTest {
...
       Offer.newBuilder()
         .setId(MesosOfferIdProto.newBuilder().setValue(UUID.randomUUID().toString))
         .setFrameworkId(FrameworkID.newBuilder().setValue("notanidframework"))
         .setSlaveId(SlaveID.newBuilder().setValue(s"slave1"))
         .setHostname("hostname")
         .setDomain(DomainInfo.newBuilder()
-          .setFaultDomain(FaultDomain.newBuilder()
-            .setRegion(RegionInfo.newBuilder().setName(faultRegion))
-            .setZone(ZoneInfo.newBuilder().setName("zone")))
+          .setFaultDomain(FaultDomainPB.newBuilder()
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala]: nit: dislike the `PB` in `FaultDomainPB`. May be `Protos.FaultDomain` or similar?

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 2]
```diff
@@ -133,7 +133,7 @@ object ResourceMatcher extends StrictLogging {
     * the reservation.
     */
   def matchResources(offer: Offer, runSpec: RunSpec, knownInstances: => Seq[Instance],
-    selector: ResourceSelector, conf: MatcherConf, schedulerPlugins: Seq[SchedulerPlugin] = Seq.empty)(implicit clock: Clock): ResourceMatchResponse = {
+    selector: ResourceSelector, conf: MatcherConf, schedulerPlugins: Seq[SchedulerPlugin] = Seq.empty, homeFaultDomain: Option[FaultDomain])(implicit clock: Clock): ResourceMatchResponse = {
```
Comment (zen-dog) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: May be give it a default of `None` and avoid passing it altogether in tests later? Makes for a lot less changed lines.

File: src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala [hunk 3]
```diff
@@ -29,13 +29,6 @@ class ActorOfferMatcher(actorRef: ActorRef, override val precedenceFor: Option[P
 
   override def toString: String = s"ActorOfferMatcher($actorRef)"
 
-  override def isInterestedIn(offer: Offer) = {
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala]: Where did this method go?

Comment (meln1k) [reply src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala]: It's true by default, and we need exactly this

Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala]: If it's `true` in the trait and we never override it - do we really need it?

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala (hunk 2 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: 👍

File: src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala [hunk 4]
```diff
@@ -18,7 +18,7 @@ import scala.concurrent.{ Future, Promise }
   * @param actorRef Reference to actor that matches offers.
   * @param precedenceFor Defines which matcher receives offers first. See [[mesosphere.marathon.core.matcher.base.OfferMatcher.precedenceFor]].
   */
-class ActorOfferMatcher(actorRef: ActorRef, override val precedenceFor: Option[PathId], homeRegion: () => Option[String])
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcher.scala]: I think the whole homeFaultDomain is not needed here anymore ... ?

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 5]
```diff
@@ -176,6 +176,12 @@ object ResourceMatcher extends StrictLogging {
 
     def portsMatchOpt: Option[PortsMatch] = PortsMatcher(runSpec, offer, selector).portsMatch
 
+    val meetsFaultDomainRequirements: Boolean = {
+      val faultDomainFields = Set(Constraints.regionField, Constraints.zoneField)
+      val hasFaultDomainConstraints = runSpec.constraints.exists(c => faultDomainFields.contains(c.getField))
+      hasFaultDomainConstraints || homeFaultDomain.map(_.region) == OfferUtil.region(offer)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: Hm, I'm not such a big fan of comparing options. How about

```scala
homeFaultDomain.exists( domain => OfferUtil.region(offer).contains(domain.region) )
```

I don't have a strong opinion though.

File: src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala (hunk 1 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala]: +1

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala (hunk 2 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: +1 this would make things a little simpler :).

File: src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala (hunk 1 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala]: Well, it will be a bit wordy: `Protos.DomainInfo.FaultDomain`

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala (hunk 2 above)
Comment (meln1k) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: I agree that it would make it simpler, but in this case I'd make it explicit to not miss a single important part. If it looks too scary, can change it for sure :)

File: src/main/scala/mesosphere/marathon/MarathonScheduler.scala [hunk 6]
```diff
@@ -148,22 +149,24 @@ class MarathonScheduler(
     }
   }
 
-  protected def updateHomeRegion(masterInfo: MasterInfo): Unit = {
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/MarathonScheduler.scala]: I would save `MasterInfo` completely providing helper methods like `getHomeRegion` to access parts of it.

File: src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala [hunk 7]
```diff
@@ -1,46 +1,18 @@
...
 import mesosphere.AkkaUnitTest
 import mesosphere.marathon.core.matcher.base.OfferMatcher.MatchedInstanceOps
 import mesosphere.marathon.test.MarathonTestHelper
 import mesosphere.mesos.protos.Implicits._
 import mesosphere.mesos.protos.OfferID
-import org.apache.mesos.Protos.DomainInfo.FaultDomain
-import org.apache.mesos.Protos.DomainInfo.FaultDomain.{ RegionInfo, ZoneInfo }
-import org.apache.mesos.Protos.{ DomainInfo, FrameworkID, Offer, SlaveID, OfferID => MesosOfferIdProto }
 
 class ActorOfferMatcherTest extends AkkaUnitTest {
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala]: I understand that we removed `isInterestedIn` method but we should still unit test the use cases below.

File: src/main/scala/mesosphere/marathon/core/matcher/manager/OfferMatcherManagerModule.scala [hunk 8]
```diff
@@ -41,6 +41,6 @@ class OfferMatcherManagerModule(
     * offers.
     */
   val globalOfferMatcherWantsOffers: Observable[Boolean] = offersWanted
-  val globalOfferMatcher: OfferMatcher = new ActorOfferMatcher(offerMatcherMultiplexer, None, homeRegion)
+  val globalOfferMatcher: OfferMatcher = new ActorOfferMatcher(offerMatcherMultiplexer, None)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/matcher/manager/OfferMatcherManagerModule.scala]: Why won't we pass the region anymore? Is this a bug?

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/matcher/manager/OfferMatcherManagerModule.scala]: I'm probably missing the obvious but it seems to me we are not matching the region anymore. Where is that done?

File: src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala (hunk 7 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/core/matcher/base/util/ActorOfferMatcherTest.scala]: Done.

File: src/main/scala/mesosphere/marathon/core/matcher/manager/OfferMatcherManagerModule.scala (hunk 8 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/core/matcher/manager/OfferMatcherManagerModule.scala]: In the resource matcher

**Comment by meln1k** [comment]
@jeschkies yep, we have
